flytekit/core/context_manager.py

```diff
@@ -89,6 +89,8 @@ class Builder(object):
         execution_date: typing.Optional[datetime] = None
         logging: Optional[_logging.Logger] = None
         task_id: typing.Optional[_identifier.Identifier] = None
+        rendered_decks: Optional[List[str]] = None
```

Comment:
At this point, I think having both `decks` and `rendered_decks` complicates this a bit too much. I prefer to have `decks` be the the same as "rendered decks" and not have two notions of decks.

If the user passes in `decks=["source_code"]`, then the "input", and "output" decks are never created. This should be simple to do in `PythonTask._write_decks`.

The tough one is the timeline deck, which is created by the context manager and not `PythonTask`. Here is a workaround I see:

- Override `TimeLineDeck.__init__` to accept a `auto_add_to_deck`. If this is `False`, then the 
- In `PythonTask._output_to_literal_map` and add the deck before `@timeit` is called and  set it to `auto_add_to_deck=True` if the user passes in "timeline" in `@task(decks=)`. Otherwise, `auto_add_to_deck` is False.
- Since `PythonTask` already added the `TimelineDeck`, the `ExecutionParameters.timeline` deck will not created another one itself.

Reply:
Does that imply we need to create a Timelinedeck before any of the `timeit` calls or decorators? Or am I missing anything? Thanks.

Reply:
But the timeline is created before we [load the task](https://github.com/flyteorg/flytekit/blob/c78142b5aa24a97694150d9a6714703c4b3d56fd/flytekit/core/python_auto_container.py#L250-L256)

Could we just keep the `Timeline` deck as default? we already show the [warning](https://github.com/flyteorg/flytekit/blob/b7de700a3a779866c895f37550b95173498d2758/flytekit/deck/deck.py#L95) in the deck if users don't install the required dependencies.

Reply:
> At this point, I think having both decks and rendered_decks complicates this a bit too much. I prefer to have decks be the the same as "rendered decks" and not have two notions of decks.

agree, I don't think we need `rendered_decks`

Reply:
> Could we just keep the Timeline deck as default? we already show the warning in the deck if users don't install the required dependencies.

In it's current state, I think it's a poor user experience to have a timeline deck that shows an error message. Personally, I do not like installing `flytekitplugins-deck-standard` because it's dependencies make the image size much bigger. (Last time I checked it was +500MB).

Can we make `TimelineDeck` not be added to `decks` by default? Concretely, I mean to change the following:

https://github.com/flyteorg/flytekit/blob/c78142b5aa24a97694150d9a6714703c4b3d56fd/flytekit/core/context_manager.py#L284-L286

to `TimeLineDeck("Timeline", auto_add_to_deck=False)`. Then in `PythonTask`, it'll manually add the timeline deck if it is in `@task(decks=[...])`.

The alternative is to have simplified timeline deck that renders a normal HTML table without the graph. I see the TimelineDeck as a debugging tool that is not useful day to day, which means I would not recommend installing the `flytekitplugins-deck-standard` by default. If there is an issue with a task, then a simple HTML table that breaks down the time spend for each step is more than enough.

Reply:
I guess simply adding auto_add_to_deck=False doesn't work because the first `timeit` call happens in `load_task`, which invokes the https://github.com/flyteorg/flytekit/blob/c78142b5aa24a97694150d9a6714703c4b3d56fd/flytekit/core/context_manager.py#L284-L286 before we can know if timeline deck is in decks, thus we may lose access to the originally created timeline_deck when we enter the execution part of the `PythonTask`. 
I guess it would be feasible as long as we make timelinedeck an attribute of `ExecutionParameters `. For instance, `ExecutionParameters._timelinedeck `, and we can save the timelinedeck upon `timeline_deck` property access, then we can later decide whether to add it to the decks in `PythonTask`.

Reply:
> before we can know if timeline deck is in decks

From reading the code, when `TimelineDeck` gets created, it automatically adds itself to decks, when it calls `super().__init__` which ends up here:

https://github.com/flyteorg/flytekit/blob/f4d894af2ebb66f4492506ebb121509ef8cb8aa3/flytekit/deck/deck.py#L58

The `auto_add_to_deck=False` will turn this off:

```python
    def __init__(self, name: str, html: Optional[str] = "", auto_add_to_deck: bool = True):
        self._name = name
        self._html = html
        if auto_add_to_deck
            FlyteContextManager.current_context().user_space_params.decks.append(self)
```

Reply:
I agree, but if I turn it off and not save it anywhere else, I think I can't retrieve the initially created timeline_deck when I want to append a new time info to the original deck. Since in the current version, the only place where the `Timelinedeck` is saved is in the `decks`. For example, when we are loading task in `load_task`, it automatically creates a timelinedeck due to a `timeit` decorator, and now we turn it off so it is not saved in the `ExecutionParams`. Later when we finally load the task module and get access to the `decks` of the task, where I finally get to know if timelinedeck should be in `ExecutionParams.decks`, I think I lost the access of the previously created timelinedeck since we turned the flag off and it was absent in my context, so I can't really append any other time info to it. In this case, when another `timeit` is invoked, it will create a new `TimelineDeck` instead of using the original one.